### ISSUE TITLE
Subscriptions fail if more then 2-3 fields selected on any table [2.4.0-beta.1]

### ISSUE THREAD

**Issue by m4ttheweric** [hasura/graphql-engine#8267]
### Version Information

Server Version: 2.4.0-beta.1

### Environment

SQL Server DB
Using 2.4.0-beta.1 Docker Image

This is super easy to repro. Use the API explorer, pick any table, and try to create a subscription with > 4 or so fields.

I have noticed some slight variation as to when a subscription will fail depending on the table.

Downgraded to 2.3.0 main release and error is not occurring. Seems limited to 2.4.0-beta.1

Here's some examples:

### Example:

(works the same way with several tables so showing this one for simplicity)

**Works with 2 fields:**
![Screen Shot 2022-03-02 at 1 35 47 PM](https://user-images.githubusercontent.com/49927862/156436680-debb7c7f-1c87-4b8f-98f7-167ae97e61b5.png)

**Fails with 3 fields:**
![Screen Shot 2022-03-02 at 1 35 58 PM](https://user-images.githubusercontent.com/49927862/156436682-f7e0f549-c0c4-4547-a167-8d20a805a1eb.png)

**Changed `subscription` to `query` and it works:**
![Screen Shot 2022-03-02 at 1 36 07 PM](https://user-images.githubusercontent.com/49927862/156436685-42e5e3d0-ec98-4a23-aa88-217c2fb086df.png)

GraphQL Query that fails

```GraphQL
subscription MySubscription {
  Users {
    Id
    UserName
    Email
  }
}
```

Error in Console:
```JSON
{
  "data": null,
  "errors": [
    {
      "extensions": {
        "path": "$",
        "code": "unexpected"
      },
      "message": "database query error"
    }
  ]
}
```
Generated SQL (taken from `Analyze` button modal in Console):

```SQL
SELECT
  [row].[result_id] AS [result_id],
  [result].[json] AS [result]
FROM
  OPENJSON(
    (
      N'' + NCHAR(91) + '' + NCHAR(91) + '' + NCHAR(34) + '00000000-0000-0000-0000-000000000000' + NCHAR(34) + ',' + NCHAR(123) + '' + NCHAR(34) + 'synthetic' + NCHAR(34) + '' + NCHAR(58) + '' + NCHAR(91) + '' + NCHAR(93) + ',' + NCHAR(34) + 'query' + NCHAR(34) + '' + NCHAR(58) + '' + NCHAR(123) + '' + NCHAR(125) + ',' + NCHAR(34) + 'session' + NCHAR(34) + '' + NCHAR(58) + '' + NCHAR(123) + '' + NCHAR(125) + '' + NCHAR(125) + '' + NCHAR(93) + '' + NCHAR(93) + ''
    )
  ) WITH (
    [result_id] UNIQUEIDENTIFIER '$[0]',
    [result_vars] NVARCHAR(MAX) '$[1]' AS JSON
  ) AS [row]
  OUTER APPLY (
    SELECT
      (
        SELECT
          ISNULL(
            (
              SELECT
                [t_OPR_SECD1].[Id] AS [Id],
                [t_OPR_SECD1].[UserName] AS [UserName],
                [t_OPR_SECD1].[Email] AS [Email]
              FROM
                [dbo].[OPR_SECD] AS [t_OPR_SECD1] FOR JSON PATH,
                INCLUDE_NULL_VALUES
            ),
            (N'' + NCHAR(91) + '' + NCHAR(93) + '')
          ) AS [root]
      ) AS [Users] FOR JSON PATH,
      INCLUDE_NULL_VALUES,
      WITHOUT_ARRAY_WRAPPER
  ) AS [result]([json]) FOR JSON PATH,
  INCLUDE_NULL_VALUES
```

I copied that SQL into Azure Data Studio and executed it without any errors. It outputted a single column with JSON. I went a step further and used `JSON.parse()` in the Chrome console to make sure the JSON string was parseable and valid and it seemed to work fine and outputted the expected object.

**Comment by dennislaviolette** [comment]
Hi @m4ttheweric - We are looking into this and I should be able to get back to you with actions by early next week.

cc: @rakeshkky

**Comment by plcplc** [comment]
Hi @m4ttheweric.

I've had a look, but I'm unable to reproduce the error you're seeing.

To compare, my setup used an MSSQL server in a docker container (as produced by the `scripts/dev.sh mssql` auxiliary development script), with a table:
```SQL
create table test_subscriptions (
  id int primary key,
  field1 varchar(max),
  field2 int,
  field3 varchar(10),
  field4 varchar(max)
)
```

Then I ran a subscription
```graphql

subscription MySubscription {
  test_subscriptions {
    field1
    field2
    field3
    field4
    id
  }
}
```
on both the `2.4.0-beta1` docker image and the tip of the `main` branch, and inserting new rows updated the subscription results as you would expect.

Seeing as you mentioned Azure, I'm wondering if there might be some critical difference in the SQL Server versions of the docker image I used and what Azure provides.

Can you elaborate a bit on your specific setup? If you can provide a reproducible setup via e.g. docker images that would be tremendously helpful :-)

**Comment by m4ttheweric** [comment]
@plcplc 

Thanks for the reply. 

Couple of high level points:

* None of this occurred or occurs currently on the previous versions of Hasura against my same setup. This only started happening when I upgraded hasura to 2.4.0. Downgrading to 2.3.0 resolves the issue.
* I'm using the standard sql server docker image. I am just using azure data studio (mac's version of SSMS) to execute queries etc. 

Here's the revenant bits from my docker compose:

```yaml 
sql-server:
      container_name: 'sql-server'
      restart: always
      volumes:
         - sql_data:/var/lib/sql-server/data
         - sql_databases:/var/opt/mssql/data
      build:
         context: .
         dockerfile: Dockerfile.SqlServer
      ports:
         - '1433:1433'
      environment:
         SA_PASSWORD: ...
         ACCEPT_EULA: 'Y'
         MSSQL_PID: 'Developer'
   postgres:
      container_name: 'hasura-postgres'
      image: postgres:12
      restart: always
      volumes:
         - postgres_data:/var/lib/postgresql/data
      environment:
         POSTGRES_PASSWORD: ...
   graphql-engine:
      container_name: 'hasura'
      image: hasura/graphql-engine:v2.4.0-beta.1
      ports:
         - '3030:8080'
      depends_on:
         - postgres
         - sql-server
      restart: always
      environment:
         HASURA_GRAPHQL_CORS_DOMAIN: '*'
         ## postgres database to store Hasura metadata
         HASURA_GRAPHQL_METADATA_DATABASE_URL: postgres://postgres:...@postgres:5432/postgres
         ## enable the console served by server
         HASURA_GRAPHQL_ENABLE_CONSOLE: 'true' # set to "false" to disable console
         ## enable debugging mode. It is recommended to disable this in production
         HASURA_GRAPHQL_DEV_MODE: 'true'
         HASURA_GRAPHQL_ENABLED_LOG_TYPES: startup, http-log, webhook-log, websocket-log, query-log
         HASURA_GRAPHQL_ADMIN_SECRET: ...
         HASURA_GRAPHQL_JWT_SECRET: '{"type": "HS256", "key": ..., "claims_namespace": "hasura_claims", "claims_format": "json" }'
         HASURA_GRAPHQL_DB_URL: 'Driver={ODBC Driver 17 for SQL Server};Server=tcp:host.docker.internal,1433;Database=answr;Uid=...;Pwd=...@;'
         ACTION_HANDLER_URL: 'http://host.docker.internal:3020'
   # if running this locally, stop this container

volumes:
   postgres_data:
   sql_data:
   sql_databases:
   expo_web:

```

**Comment by m4ttheweric** [comment]
Ok so I am seeing some behavior that might be worth noting:

Back to the same table I started with:

```graphql
subscription MySubscription {
  Users {
    PasswordHash
  }
}
```

This one fails with the same error.

But if I add a where statement like this:

```graphql
subscription MySubscription {
  Users(where: {Id: {_lt: 100}}) {
    PasswordHash
  }
}
```

Then it works.

So, then I went for the `limit` option and found that anything higher than a limit of 23 works. But after that it fails.

So, this works:

```graphql
subscription MySubscription {
  Users(limit: 23) {
    PasswordHash
  }
}
```

If I change the order by then it allows for 19. So, my first thought was perhaps there was some bad data, so I switched to query and identified the rows that it was tripping on. I changed my subscription to target only those rows by Id with the same fields selected and it worked.

My best guess right now is that it's exceeding some character limit in some kind of string initialization? I really don't know.  But it seems like when you select too many rows of these long strings (our password hashes are about 84 characters long), then it fails.

**Comment by plcplc** [comment]
Thanks for that. 

I was able to reproduce the issue locally with the data:

```SQL

insert into 
  test_subscriptions(id, field1)
  values
  (10, '1234567890'),
  (20, '12345678901234567890'),
  (30, '123456789012345678901234567890'),
  (40, '1234567890123456789012345678901234567890'),
  (50, '12345678901234567890123456789012345678901234567890'),
  (60, '123456789012345678901234567890123456789012345678901234567890'),
  (70, '1234567890123456789012345678901234567890123456789012345678901234567890'),
  (80, '12345678901234567890123456789012345678901234567890123456789012345678901234567890'),
  (90, '123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890'),
  (100, '1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890')
  (101, '1234567890'),
  (201, '12345678901234567890'),
  (301, '123456789012345678901234567890'),
  (401, '1234567890123456789012345678901234567890'),
  (501, '12345678901234567890123456789012345678901234567890'),
  (601, '123456789012345678901234567890123456789012345678901234567890'),
  (701, '1234567890123456789012345678901234567890123456789012345678901234567890'),
  (801, '12345678901234567890123456789012345678901234567890123456789012345678901234567890'),
  (901, '123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890'),
  (1001, '1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890')
 ```

It does seem to be related to the size of the response in some way.

**Comment by plcplc** [comment]
Okay, my findings so far seems indicate that we bump into [MSSQL splitting up results across multiple rows](https://stackoverflow.com/questions/54973536/for-json-path-results-in-ssms-truncated-to-2033-characters) (see [docs](https://docs.microsoft.com/en-us/sql/relational-databases/json/use-for-json-output-in-sql-server-and-in-client-apps-sql-server?view=sql-server-2017#use-for-json-output-in-a-c-client-app)) when the top-level query is `SELECT .. FOR JSON` (which the subscription query has).

This could well have been introduced in a recent refactor of the SQL generation code we did.

At least I can see that if I take the result splitting into account the subscription query starts working again, so I'm optimistic we'll have a fix for that shortly.

**Comment by m4ttheweric** [comment]
I'm glad we could work to get this repro'd and resolved! Thanks for your engagement on this.

**Comment by tirumaraiselvan** [comment]
This has been released in `v2.4.0-beta.3`
